File: angular/studentmanagerapp/src/app/services/announcement.service.ts
import { HttpClient, HttpHeaders } from '@angular/common/http';
import { Injectable } from '@angular/core';
import { Observable } from 'rxjs';
import { AddAnnouncementInterface } from '../interfaces/add-announcement-interface';
import { EditAnnouncementInterface } from '../interfaces/edit-announcement-interface';
import { Announcement } from '../models/announcement.model';

@Injectable({
  providedIn: 'root'
})
export class AnnouncementService {
  private apiUrl: string = 'http://localhost:8080/api';
  headers = new HttpHeaders().set('Content-Type', 'application/json');

  constructor(private http: HttpClient) { }

  getUserAnnouncements(userId: number): Observable<Announcement[]> {
    const url: string = `${this.apiUrl}/announcements/user/${userId}`;

    return this.http.get<Announcement[]>(url, { headers: this.headers });
  }

  getCourseAnnouncements(courseId: number): Observable<Announcement[]> {
    const url: string = `${this.apiUrl}/announcements/course/${courseId}`;

    return this.http.get<Announcement[]>(url, { headers: this.headers });
  }

  editAnnouncement(announcementId: number, body: EditAnnouncementInterface): Observable<Announcement> {
    const url: string = `${this.apiUrl}/announcement/${announcementId}`;

    return this.http.patch<Announcement>(url, body);
  }

  saveAnnouncement(body: AddAnnouncementInterface): Observable<Announcement> {
    const url: string = `${this.apiUrl}/announcement`;

    return this.http.post<Announcement>(url, body, { headers: this.headers });
  }

  deleteAnnouncement(announcementId: number): Observable<any> {
    const url: string = `${this.apiUrl}/announcement/${announcementId}`;

    return this.http.delete(url, { responseType: 'text' });
  }
}

File: src/main/java/sda/studentmanagement/studentmanager/services/AnnouncementService.java
package sda.studentmanagement.studentmanager.services;

import sda.studentmanagement.studentmanager.domain.Announcement;
import sda.studentmanagement.studentmanager.dto.AddAnnouncementFormDto;
import sda.studentmanagement.studentmanager.dto.EditAnnouncementFormDto;

import javax.validation.ConstraintViolationException;
import java.util.List;

public interface AnnouncementService {
    Announcement getAnnouncement(long announcementId);

    List<Announcement> getUserAnnouncements(long userId);

    List<Announcement> getCourseAnnouncements(long courseId);

    Announcement saveAnnouncement(AddAnnouncementFormDto announcement) throws ConstraintViolationException;

    Announcement editAnnouncement(long announcementId, EditAnnouncementFormDto announcementForm);

    void deleteAnnouncement(long announcementId);
}
